[Feat][gpt-oss] Add token details to continuous usage stats

NagyGeorge opened the pull request:
<!-- markdownlint-disable -->

## Changes

  - Added CompletionTokenUsageInfo class to track reasoning tokens separately
  - Extended UsageInfo with completion_tokens_details field for reasoning token tracking
  - Implemented reasoning token counting for Harmony/GPT-OSS models based on analysis channels and tool calls
  - Added prompt_tokens_details to continuous usage stats for both chat and completion endpoints
  - Included cached token information when enable_prompt_tokens_details is enabled

## Purpose
  - Cost tracking: Users can accurately track token usage including reasoning tokens for billing purposes
  - Performance monitoring: Cached token information helps identify cache hit rates

## Test Plan
- CI Pipeline

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

NagyGeorge committed:
[Feat][gpt-oss] Add token details to continuous usage stats for streaming

- Add CompletionTokenUsageInfo class to track reasoning tokens
  - Add completion_tokens_details field to UsageInfo for reasoning token
  details
  - Track reasoning tokens for Harmony/gpt_oss completion_tokens_details
  - Include prompt_tokens_details in continuous usage for chat and
    completion details
  - Include completion_tokens_details in continuous usage for Harmony
    models

Signed-off-by: George Nagy II <george.nagy0969@gmail.com>

NagyGeorge committed:
refactor: centralize prompt token usage info logic

- Add _create_prompt_token_usage_info() helper method to OpenAIServing base class                                                                                             │
- Eliminate code duplication across serving_chat.py and serving_completion.py

Signed-off-by: George Nagy II <george.nagy0969@gmail.com>

NagyGeorge commented:
@heheda12345

NagyGeorge commented:
/gemini review

NagyGeorge committed:
refactor: improve token caching logic and simplify tool call check

  - Fix cached token condition to explicitly check for None instead of falsy value
  - Simplify is_tool_call check by using tuple with startswith() method

Signed-off-by: George Nagy II <george.nagy0969@gmail.com>

yeqcharlotte reviewed (changes requested):
also i don't think these logic got automatic CI coverage. better look at existing tests and see what's missing :)

NagyGeorge commented on vllm/entrypoints/openai/serving_chat.py:
```diff
@@ -667,6 +681,18 @@ async def chat_completion_stream_generator(
+                            # Track reasoning tokens based on current
+                            # channel/recipient. Similar to PR #23460
+                            # implementation in context.py
+                            is_analysis = (
+                                harmony_parser.current_channel == "analysis")
+                            is_tool_call = (
+                                harmony_parser.current_recipient is not None
+                                and
+                                harmony_parser.current_recipient.startswith(
+                                    ("python", "browser.")))
```
Thanks for the feedback I will work on it!

NagyGeorge committed:
refactor: enhance reasoning token detection and eliminate duplication

  - Create shared is_reasoning_token() utility in harmony_utils.py
  - Expand reasoning token support from python/browser only to ALL recipients
  - Add commentary channel support alongside existing analysis channel
  - Fix streaming chat final usage to include aggregated reasoning tokens
  - Add reasoning token calculation to non-streaming chat responses
  - Eliminate code duplication between serving_chat.py and context.py
  - Fix streaming context reasoning token counting to process per-token instead of per-output
  - Add comprehensive test coverage for all recipient types and channels
  - Update mock fixtures to prevent false positives in testing

Signed-off-by: George Nagy II <george.nagy0969@gmail.com>

yeqcharlotte commented:
thanks for the change @NagyGeorge! can you include some e2e results? like given a prompt, what's the result token usage details after the change? and double check if serving_chat unit test cover the counting logic?

NagyGeorge committed:
tests: add unit tests for token usage details

- Add test_serving_chat_token_details_prompt_tokens
  - Add test_serving_chat_token_details_disabled
  - Add test_serving_chat_completion_tokens_details_reasoning
  - Add test_serving_chat_streaming_token_details

Signed-off-by: George Nagy II <george.nagy0969@gmail.com>

yeqcharlotte commented:
@NagyGeorge how about giving a try using bf16 gptoss using lmsys/gpt-oss-20b-bf16

NagyGeorge committed:
fix: synchronize messages in streaming context

- add message synchronization in streamingharmonycontext
  - Ensure parser messages are properly extended to context._messages
  - Fix test to verify synchronization occurs after message completion

Signed-off-by: George Nagy II <george.nagy0969@gmail.com>

yeqcharlotte commented:
> @yeqcharlotte
> 
> bfloat16 still requires compute 8.0. Switched to half precision and failed due to to FlexAttention not supporting sliding window. Installed FlashInfer and tried with various GPU utilization settings and all end with GPU OOM. I could be doing something wrong here but I just don't believe my GPU can run the 20B model. Working on the unit tests, of course e2e validation would be great and I will see if I can source a way to test it one way or another.

i see. feel free to send me a small test client and launch code and i can run them on my side.
